std/math/emulated/field_test.go

```diff
@@ -0,0 +1,388 @@
+	res.Mod(res, q)
+	return x1, x2, x3, x4, x5, x6, res
+}
+
+type WrapperCircuit struct {
+	X1, X2, X3, X4, X5, X6 frontend.Variable
+	Res                    frontend.Variable
+}
+
+func (c *WrapperCircuit) Define(api frontend.API) error {
+	// compute x1^3 + 5*x2 + (x3-x4) / (x5+x6)
+	x13 := api.Mul(c.X1, c.X1, c.X1)
+	fx2 := api.Mul(5, c.X2)
+	nom := api.Sub(c.X3, c.X4)
+	denom := api.Add(c.X5, c.X6)
+	free := api.Div(nom, denom)
+	res := api.Add(x13, fx2, free)
+	api.AssertIsEqual(res, c.Res)
+	return nil
+}
+
+func TestTestEngineWrapper(t *testing.T) {
+	assert := test.NewAssert(t)
+
+	circuit := WrapperCircuit{
+		X1:  NewElement[Secp256k1](nil),
+		X2:  NewElement[Secp256k1](nil),
+		X3:  NewElement[Secp256k1](nil),
+		X4:  NewElement[Secp256k1](nil),
+		X5:  NewElement[Secp256k1](nil),
+		X6:  NewElement[Secp256k1](nil),
+		Res: NewElement[Secp256k1](nil),
+	}
+
+	x1, x2, x3, x4, x5, x6, res := witnessData(Secp256k1{}.Modulus())
+	witness := WrapperCircuit{
+		X1:  NewElement[Secp256k1](x1),
+		X2:  NewElement[Secp256k1](x2),
+		X3:  NewElement[Secp256k1](x3),
+		X4:  NewElement[Secp256k1](x4),
+		X5:  NewElement[Secp256k1](x5),
+		X6:  NewElement[Secp256k1](x6),
+		Res: NewElement[Secp256k1](res),
+	}
+	wrapperOpt := test.WithApiWrapper(func(api frontend.API) frontend.API {
+		napi, err := NewField[Secp256k1](api)
+		assert.NoError(err)
+		return napi
+	})
+	err := test.IsSolved(&circuit, &witness, testCurve.ScalarField(), wrapperOpt)
+	assert.NoError(err)
+}
+
+func TestCompilerWrapper(t *testing.T) {
+	assert := test.NewAssert(t)
+	circuit := WrapperCircuit{
+		X1:  NewElement[Secp256k1](nil),
+		X2:  NewElement[Secp256k1](nil),
+		X3:  NewElement[Secp256k1](nil),
+		X4:  NewElement[Secp256k1](nil),
+		X5:  NewElement[Secp256k1](nil),
+		X6:  NewElement[Secp256k1](nil),
+		Res: NewElement[Secp256k1](nil),
+	}
+
+	x1, x2, x3, x4, x5, x6, res := witnessData(Secp256k1{}.Modulus())
+	witness := WrapperCircuit{
+		X1:  NewElement[Secp256k1](x1),
+		X2:  NewElement[Secp256k1](x2),
+		X3:  NewElement[Secp256k1](x3),
+		X4:  NewElement[Secp256k1](x4),
+		X5:  NewElement[Secp256k1](x5),
+		X6:  NewElement[Secp256k1](x6),
+		Res: NewElement[Secp256k1](res),
+	}
+	ccs, err := frontend.Compile(testCurve.ScalarField(), r1cs.NewBuilder, &circuit, frontend.WithBuilderWrapper(builderWrapper[Secp256k1]()))
+	assert.NoError(err)
+	t.Log(ccs.GetNbConstraints())
+	// TODO: create proof
+	_ = witness
+}
+
+func TestIntegrationApi(t *testing.T) {
+	assert := test.NewAssert(t)
+	wrapperOpt := test.WithApiWrapper(func(api frontend.API) frontend.API {
+		napi, err := NewField[Secp256k1](api)
+		assert.NoError(err)
+		return napi
+	})
+	keys := make([]string, 0, len(circuits.Circuits))
+	for k := range circuits.Circuits {
+		keys = append(keys, k)
+	}
+	sort.Strings(keys)
+
+	for i := range keys {
+		name := keys[i]
+		if name == "inv" || name == "div" || name == "cmp" {
+			// TODO @gbotrel thes don't pass when we use emulated field modulus != snark field
```

Comment:
Hmm. I was at some point fixing these failing tests. I'll try to pop the stash. I think inv and div were easy but there was some strange problem with cmp. I'll  see.